### PR TITLE
Reification hooks and tests

### PR DESCRIPTION
**Pull request by palimarrao**
With this PR, construct clauses can be fully expressed via the `Statement` struct.

### CODE REVIEW

File: bql/grammar/grammar_test.go [hunk 1]
```diff
@@ -412,3 +412,85 @@ func TestSemanticStatementGraphClausesLengthCorrectness(t *testing.T) {
...
+			                      "predicate_2"@[] ?o2.
+			                   ?s "predicate_3"@[] ?o3} into ?a from ?b where {?s "old_predicate_1"@[,] ?o1.
+					                                                   ?s "old_predicate_2"@[,] ?o2.
+											   ?s "old_predicate_3"@[,] ?o3};`,
+			want:  2,
+		},
+	}
+	p, err := NewParser(SemanticBQL())
+	if err != nil {
+		t.Errorf("grammar.NewParser: should have produced a valid BQL parser, %v", err)
```
Comment (xllora) [inline bql/grammar/grammar_test.go]: grammar.NewParser: **S**hould have produced a valid BQL parser

File: bql/grammar/grammar_test.go [hunk 2]
```diff
@@ -412,3 +412,85 @@ func TestSemanticStatementGraphClausesLengthCorrectness(t *testing.T) {
...
+	}
+	for _, entry := range table {
+		st := &semantic.Statement{}
+		if err := p.Parse(NewLLk(entry.query, 1), st); err != nil {
+			t.Errorf("Parser.consume: failed to accept valid semantic entry %q", entry.query)
```
Comment (xllora) [inline bql/grammar/grammar_test.go]: Parser.consume: **F**ailed to accept valid semantic entry %q

File: bql/grammar/grammar_test.go [hunk 3]
```diff
@@ -412,3 +412,85 @@ func TestSemanticStatementGraphClausesLengthCorrectness(t *testing.T) {
...
+			                      "predicate_2"@[] ?o2;
+			                      "predicate_3"@[] ?o3.
+				           ?s1 "predicate_1"@[] ?o1;
+				               "predicate_2"@[] ?o2;
+				               "predicate_3"@[] ?o3} into ?a from ?b where {?s "old_predicate_1"@[,] ?o1.
+					                                                    ?s "old_predicate_2"@[,] ?o2.
+											    ?s1 "old_predicate_3"@[,] ?o3};`,
+			want_one:  2,
+			want_two:  2,
+		},
```
Comment (xllora) [inline bql/grammar/grammar_test.go]: Please repeat the tests using temporal predicates.

Comment (palimarrao) [inline bql/grammar/grammar_test.go]: Done.

File: bql/grammar/grammar_test.go (hunk 1 above)
Comment (palimarrao) [inline bql/grammar/grammar_test.go]: Done.

File: bql/grammar/grammar_test.go [hunk 4]
```diff
@@ -400,15 +400,131 @@ func TestSemanticStatementGraphClausesLengthCorrectness(t *testing.T) {
...
+		{
+			query: `construct {?s "predicate_1"@[] ?o1;
+			                      "predicate_2"@[] ?o2;
+			                      "predicate_3"@[?t] ?o3.
+				           ?s1 "predicate_1"@[] ?o1;
+				               "predicate_2"@[] ?o2;
+				               "predicate_3"@[?t] ?o3} into ?a from ?b where {?s "old_predicate_1"@[,] ?o1.
+											      ?s "old_predicate_2"@[,] ?o2.
+											      ?s1 "old_predicate_3"@[?t] ?o3};`,
```
Comment (xllora) [inline bql/grammar/grammar_test.go]: ?s1 "old_predicate_3"@[?t] ?o3

This is not a valid statement. The valid representation is by using the AT binding keyword. See the example below.

?s1 "old_predicate_3"@[,] AT ?t ?o3

Comment (palimarrao) [inline bql/grammar/grammar_test.go]: I have fixed this example.

**Comment by xllora** [comment]
Thanks!

**Pull request merged**